File: projects/08-message-broker/src/broker.rs
//! The broker: the top-level owner that maps topic names to topics and holds the
//! consumer-group coordinator. Plumbing/wiring — the interesting behaviour lives
//! in the verticals it composes (`Topic` → V3, `Log`/`Index` → V1/V2,
//! `GroupCoordinator` → V4).
//!
//! On-disk layout under `DATA_DIR`:
//! ```text
//! <data_dir>/
//!   topics/<topic>/<partition>/{…}.log + {…}.index   ← V1 + V2
//!   groups/…                                          ← V4 committed offsets
//! ```

use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::Arc;

use tokio::sync::RwLock;

use crate::error::AppError;
use crate::group::GroupCoordinator;
use crate::log::LogConfig;
use crate::topic::Topic;

/// A lightweight view of a topic for the list endpoint.
#[derive(Debug, Clone, serde::Serialize)]
pub struct TopicInfo {
    pub name: String,
    pub partitions: usize,
}

/// The broker. Cloneable via `Arc` into every request handler.
pub struct Broker {
    topics_dir: PathBuf,
    config: LogConfig,
    default_partitions: u32,
    topics: RwLock<HashMap<String, Arc<Topic>>>,
    groups: Arc<GroupCoordinator>,
}

impl Broker {
    /// Open the broker under `data_dir`, reloading any topics already on disk.
    pub fn open(
        data_dir: impl Into<PathBuf>,
        config: LogConfig,
        default_partitions: u32,
    ) -> Result<Arc<Self>, AppError> {
        let data_dir = data_dir.into();
        let topics_dir = data_dir.join("topics");
        std::fs::create_dir_all(&topics_dir)?;
        let groups = GroupCoordinator::open(data_dir.join("groups"))?;

        // Reload existing topics: each subdirectory of topics/ is a topic. Their
        // logs recover their own offsets on open (V1 recovery).
        let mut topics = HashMap::new();
        for entry in std::fs::read_dir(&topics_dir)? {
            let entry = entry?;
            if !entry.path().is_dir() {
                continue;
            }
            if let Some(name) = entry.file_name().to_str() {
                let topic = Topic::open(&topics_dir, name, config)?;
                topics.insert(name.to_string(), topic);
            }
        }

        Ok(Arc::new(Self {
            topics_dir,
            config,
            default_partitions,
            topics: RwLock::new(topics),
            groups,
        }))
    }

    /// The consumer-group coordinator (V4).
    pub fn groups(&self) -> &Arc<GroupCoordinator> {
        &self.groups
    }

    /// Create a topic. `partitions` of `None` uses the configured default.
    pub async fn create_topic(
        &self,
        name: &str,
        partitions: Option<u32>,
    ) -> Result<Arc<Topic>, AppError> {
        validate_topic_name(name)?;
        let mut topics = self.topics.write().await;
        if topics.contains_key(name) {
            return Err(AppError::TopicAlreadyExists);
        }
        let n = partitions.unwrap_or(self.default_partitions);
        let topic = Topic::create(&self.topics_dir, name, n, self.config)?;
        topics.insert(name.to_string(), topic.clone());
        Ok(topic)
    }

    /// Look up a topic by name.
    pub async fn topic(&self, name: &str) -> Result<Arc<Topic>, AppError> {
        self.topics
            .read()
            .await
            .get(name)
            .cloned()
            .ok_or(AppError::UnknownTopic)
    }

    /// List all topics (for `GET /topics`).
    pub async fn list_topics(&self) -> Vec<TopicInfo> {
        self.topics
            .read()
            .await
            .values()
            .map(|t| TopicInfo {
                name: t.name().to_string(),
                partitions: t.partition_count(),
            })
            .collect()
    }
}

/// Reject topic names that would escape the data dir or make illegal paths — the
/// name becomes a directory (see the security horizontal item).
fn validate_topic_name(name: &str) -> Result<(), AppError> {
    let ok = !name.is_empty()
        && name.len() <= 255
        && name
            .chars()
            .all(|c| c.is_ascii_alphanumeric() || matches!(c, '-' | '_' | '.'))
        && name != "."
        && name != "..";
    if ok {
        Ok(())
    } else {
        Err(AppError::InvalidRequest(format!(
            "illegal topic name: {name:?}"
        )))
    }
}

File: projects/08-message-broker/src/message_broker/broker.py
"""The broker: the top-level owner that maps topic names to topics and holds the
consumer-group coordinator.

Plumbing/wiring — the interesting behaviour lives in the verticals it composes
(`Topic` -> V3, `Log`/`Index` -> V1/V2, `GroupCoordinator` -> V4).

On-disk layout under `data_dir`::

    <data_dir>/
      topics/<topic>/<partition>/{...}.log + {...}.index   <- V1 + V2
      groups/...                                            <- V4 committed offsets

Note where the blocking work goes. Creating a topic makes directories and lays
down files, which is real filesystem I/O; it happens inside a request handler, so
it is pushed to a worker thread with `asyncio.to_thread`. It is a rare admin call
and you could get away with blocking the loop for it — but the scaffold does it
properly on purpose, because the habit is what transfers, and because "it's only
a few milliseconds" is how a p99 dies.
"""

from __future__ import annotations

import asyncio
from pathlib import Path

import structlog

from .errors import InvalidRequest, TopicAlreadyExists, UnknownTopic
from .group import GroupCoordinator
from .log import LogConfig
from .topic import Topic

__all__ = ["Broker", "validate_topic_name"]

logger = structlog.get_logger(__name__)

# A topic name becomes a directory name under `data_dir/topics/`, so it is
# attacker-controlled input that reaches the filesystem. 255 is the usual
# per-component limit on ext4/APFS/NTFS.
MAX_TOPIC_NAME_LEN = 255
_TOPIC_NAME_EXTRA = frozenset("-_.")


class Broker:
    """The broker. Constructed once at startup and shared by every handler."""

    def __init__(
        self,
        topics_dir: Path,
        config: LogConfig,
        default_partitions: int,
        topics: dict[str, Topic],
        groups: GroupCoordinator,
    ) -> None:
        self._topics_dir = topics_dir
        self._config = config
        self._default_partitions = default_partitions
        self._topics = topics
        self._groups = groups
        # Guards the topics map against two concurrent creates of the same name.
        # Without it, both would pass the "does it exist?" check and the second
        # would clobber the first's Topic object while its files sat on disk.
        self._lock = asyncio.Lock()

    @classmethod
    def open(cls, data_dir: Path, config: LogConfig, default_partitions: int) -> Broker:
        """Open the broker under `data_dir`, reloading any topics already on
        disk.

        Synchronous, and that is deliberate: this runs once at startup before the
        loop is serving anything, so there is no request to stall. Reloading is
        also the moment every partition's log recovers its own offsets (V1
        recovery), which is why a restart is the cheapest test you have.
        """
        topics_dir = data_dir / "topics"
        topics_dir.mkdir(parents=True, exist_ok=True)
        groups = GroupCoordinator.open(data_dir / "groups")

        topics: dict[str, Topic] = {}
        for entry in sorted(topics_dir.iterdir()):
            if entry.is_dir():
                topics[entry.name] = Topic.open(topics_dir, entry.name, config)
        return cls(topics_dir, config, default_partitions, topics, groups)

    @property
    def groups(self) -> GroupCoordinator:
        """The consumer-group coordinator (V4)."""
        return self._groups

    @property
    def default_partitions(self) -> int:
        return self._default_partitions

    async def create_topic(self, name: str, partitions: int | None = None) -> Topic:
        """Create a topic. `partitions=None` uses the configured default."""
        validate_topic_name(name)
        count = self._default_partitions if partitions is None else partitions
        async with self._lock:
            if name in self._topics:
                raise TopicAlreadyExists()
            topic = await asyncio.to_thread(
                Topic.create, self._topics_dir, name, count, self._config
            )
            self._topics[name] = topic
        logger.info("topic created", topic=name, partitions=topic.partition_count)
        return topic

    def topic(self, name: str) -> Topic:
        """Look up a topic by name.

        No lock and no `await`: a dict read is atomic on one event loop, and
        making the hot fetch path wait behind a topic creation would be an
        unforced error.
        """
        try:
            return self._topics[name]
        except KeyError:
            raise UnknownTopic() from None

    def list_topics(self) -> list[Topic]:
        """Every topic, name-ordered (for `GET /topics`)."""
        return [self._topics[name] for name in sorted(self._topics)]

    async def close(self) -> None:
        """Durably flush everything on shutdown.

        Called from the lifespan on SIGTERM, *after* in-flight requests have
        drained, so nothing is mid-append when the fsyncs happen. Until V1 and V4
        exist this raises `NotImplementedError`, which `main.py` catches — an
        unbuilt flush must not turn a clean shutdown into a crash.
        """
        for topic in self._topics.values():
            await topic.flush()
        await self._groups.flush()


def validate_topic_name(name: str) -> None:
    """Reject topic names that would escape the data dir or make illegal paths.

    An allowlist, not a denylist: the name becomes a directory, and a denylist of
    "bad" sequences loses to the next encoding trick. Alphanumerics plus `-_.`
    can never contain a separator or a traversal, which makes the guarantee
    structural instead of a list of things someone thought of.

    This is the security-horizontal item, and it wants its own test.
    """
    ok = (
        0 < len(name) <= MAX_TOPIC_NAME_LEN
        and all(c.isascii() and (c.isalnum() or c in _TOPIC_NAME_EXTRA) for c in name)
        # `.` and `..` pass the charset check and are exactly the traversal you
        # were guarding against.
        and name not in {".", ".."}
    )
    if not ok:
        raise InvalidRequest(f"illegal topic name: {name!r}")
